invoke/runners.py

```diff
@@ -322,8 +322,22 @@ def get():
                     # Can't use six.b because that just assumes latin-1 :(
                     data = data.encode(self.encoding)
                 yield data
-        # Decode stream using our generator & requested encoding
-        for data in codecs.iterdecode(get(), self.encoding, errors='replace'):
+        if not hasattr(output, 'encoding') or self.encoding == output.encoding:
+            # No need in re-encoding if output is a BytesIO or CarbonCopy, or
+            # source and target encodings are the same
+            data_stream = get()
+        else:
+            # Re-encode from user requested encoding to the output encoding
+            data_stream = codecs.iterencode(
+                codecs.iterdecode(
+                    get(),
+                    self.encoding,
+                    errors='replace'
+                ),
+                output.encoding,
+                errors='replace'
+            )
+        for data in data_stream:
```

Comment:
Regarding (3):
1. There is no "text" mode on Linux and OS X, write will encode unicode to bytes under the hood if we don't, so it is usually advised to pass bytes when you write or send anything in Python
2. In tests, it is used CarbonCopy, which is based on BytesIO (**bytes**)

I changed the code because it was buggy. It was just a luck that it used to work. There *was* the sort of error that you referred to in your first comment.